### ISSUE TITLE
local_user unreliable on Mac OS X

### ISSUE THREAD

**Issue by felixbuenemann** [capistrano/capistrano#1520]
Capitrano's `local_user` uses ruby's `Etc.getlogin` under the hood. Unfortunately that's not reliable on at least Mac OS X, because it is a global setting that returns whatever the last call to setlogin(2) was for the current login shell (quote from `man 2 setlogin` on Mac OS X 10.11.2):

> The getlogin() routine returns the login name of the user associated with the current session, as previously set by setlogin().  The name is normally associated with a login shell at the time a session is created, and is inherited by all processes descended from the login shell.  (This is true even if some of those processes assume another user ID, for example when su(1) is used.)

This means that if some background process started from my login session calls setlogin, the next time getlogin is called it will return that value shared across all processes that descend from the login shell.

**TL;DR:** Etc.getlogin might return names like `_mbsetupuser` or `_spotlight` and cannot be trusted.

Instead we could be relying on either `ENV['USER']` and falling back to `Etc.getlogin` if it's not set – in fact ruby's getlogin uses ENV['USER'], if the OS does not implement the getlogin syscall. Or we could use `Etc.getpwuid.name` which looks up the username from /etc/passwd given the current processes' uid obtained via getuid(2).

I'm not sure if capistrano supports Windows, if it does, we could use ENV['USERNAME'] on it.

**Comment by leehambley** [comment]
Hi Felix, can you can prepare a PR for relying on `ENV['USER']` I believe on most platforms that's going to be mostly reliable. I'm a little shocked that `Etc.getlogin` is so broken on OSX, but - seems to be the current trend.

We do support Windows, so `Env['User'] || Env['Username']` would cover the bases

**Comment by felixbuenemann** [comment]
I'm afraid that's got less to do with OS X, than it's BSD kernel. I looked at the manpage for FreeBSD and it has pretty much the identical text.

I'll come up with a PR to use the environment variables shortly.
